azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/ReactorDispatcher.java

```diff
@@ -135,26 +135,26 @@ public void run(Selectable selectable) {
     private final class CloseHandler implements Callback {
         @Override
         public void run(Selectable selectable) {
+            workScheduler.run(null);
```

Comment:
>workScheduler.run(null); [](start = 12, length = 24)

why is the order changed ?
The rationale behind old order is that:
after closing the channel - no more writes are allowed, so, it is safer to drain all the tasks.

If tasks are drained - and then, if the channel is closed - won't new tasks be added in between?

Reply:
For the most part, close handler will be invoked as part of recreating the reactor due to an error and scheduling a new task will be put into a new reactor and tasks in the old reactor will be drained fully before close handler is invoked. Please note that workScheduler.run reads the pipe first before draining tasks and closed-channel error will be thrown if the channel was already closed.

Reply:
"For the most part, close handler will be invoked as part of recreating the reactor due to an error"
This statement is wrong! Close handler is invoked every time reactor is free'd - which means this is invoked - when MessagingFactory.close() call is invoked. 

I still did not get your point on, draining tasks and closed channel.
Let me explain this: 
signaling work queue (in the invoke & invoke with delay methods above) is intended to run on a different thread to that of CloseHandler (which runs on Reactor thread). That is the functionality that ReactorDispatcher offers.
So, in this PR, after reversing the order to workScheduler.run() and channel.close(), we are introducing a chance for a race condition - where new task is inserted into the ReactorDispatcher queue - **after** workScheduler.run() finishes and **before** channel.close() is invoked. 
Whereas, the original code handles this case very well.

So, my main question is - please explain what was the motivation - behind reversing the order; was there a customer reported incident - which somehow lead us to this conclusion?